File: MonotonicStack/1081_Smallest_Subsequence_of_Distinct_Characters/1081_Smallest_Subsequence_of_Distinct_Characters.py
class Solution(object):
    def smallestSubsequence(self, s):
        """
        :type s: str
        :rtype: str
        1081. Smallest Subsequence of Distinct Characters
        Medium
        Given a string s, return the lexicographically smallest subsequence
        of s that contains all the distinct characters of s exactly once.

        Example 1:
        Input: s = "bcabc"
        Output: "abc"

        Example 2:
        Input: s = "cbacdcbc"
        Output: "acdb"
        """

        # first use a hash to count characters with number of occurrences
        # then use a mono stack to put chars in ascending order as much as you can
        # you cannot do completetely, else you would simply
        # do a sort on ther set of chars
        
        hash = {}
        for c in s:
            hash[c] = hash.get(c, 0) + 1

        stack = []

        for c in s:
            hash[c] -= 1    # Decrement the count as we are using this character now
            if c in stack:  # Skip if the character is already in the result stack
                continue

            # Check stack is not empty and conditions for pop
            while stack and c < stack[-1] and hash[stack[-1]] > 0:
                stack.pop()

            stack.append(c)
        
        return "".join(stack)

File: MonotonicStack/1081_Smallest_Subsequence_of_Distinct_Characters/1081_Smallest_Subsequence_of_Distinct_Characters.cpp
class Solution {
public:
    string smallestSubsequence(string s) {
        // Using unordered_map as a substitute for Python's dict
        std::unordered_map<char, int> hash;

        // Populating the hash map
        for (char c : s) {
            hash[c] = hash.find(c) != hash.end() ? hash[c] + 1 : 1;
        }

        // Using vector as a substitute for Python's list (acting as a stack)
        std::vector<char> stack;

        for (char c : s) {
            hash[c]--; // Decrement the count as we are using this character now

            // Skip if the character is already in the result stack
            if (std::find(stack.begin(), stack.end(), c) != stack.end()) {
                continue;
            }

            // Check stack is not empty and conditions for pop
            while (!stack.empty() && c < stack.back() && hash[stack.back()] > 0) {
                stack.pop_back();
            }

            stack.push_back(c);
        }

        // Converting the vector to a string
        return std::string(stack.begin(), stack.end());        
    }
};
